### ISSUE TITLE
qwen4_exp loader OOM-kills on a 180 GB-RAM box: the whole artifact is pre-materialized on host before the first H2D

### ISSUE THREAD

**Issue by avifenesh** [avifenesh/memra#48]
## Defect

`qwen4exp_real_gate` loading the real NVFP4 artifact (174 GB mint incl. the 102 GB bf16
n-gram table) is **OOM-killed by the host kernel at anon-RSS 179.7 GB on a 180 GB-RAM
box**, three times in a row, with 64 GB of swap untouched (pinned/anon growth outruns
reclaim):

```
Out of memory: Killed process (qwen4exp_real_g) total-vm:575237472kB, anon-rss:179739000kB
```

The lane box that first brought this artifact up had 499 GB of RAM, so the defect was
never exposed there. 180 GB is the cheapest 2-card class on every provider we rent from,
so this is a hardware-cost defect, not a comfort one.

## Mechanism

`crates/memra-engine/src/qwen4exp_gpu.rs`:

- `load_from_dir_with` → `read_checkpoint_with(dir, opts)` materializes the **entire**
  `LoadedCheckpoint` on host: every layer's expert bank is staged into a `Vec` and parked
  in `LoadedCheckpoint::banks` (~72 GB as-stored NVFP4 across 48 layers), alongside the
  102 GB n-gram table and ~20 GB of trunk f32 in `ReferenceWeights`.
- Only **then** does `from_loaded_checkpoint_dual` consume `banks` and upload.

Peak host = the whole artifact ≈ 175-180 GB + margin. The n-gram table is host-resident
**by design** (gathered 16 rows/token, host math) and is not the excess; the expert banks
are device-resident as-stored and have no reason to all sit on host at once.

## Fix shape

Stream the expert banks: record per-layer read addresses during the contract walk, then
read one layer's bank off the safetensors mmap inside the consume loop, upload it, drop
it, next layer. Target peak host ≈ n-gram table + one layer's bank staging + trunk
transients.

This is a **memory-ordering** change, not a numerics change: the uploaded device bytes
must be byte-identical to today's, proven per-bank on a tiny fixture and by greedy-chain
identity against the banked real-checkpoint goldens.

## Acceptance

- old binary dies / new binary loads + passes `qwen4exp_real_gate`'s real-checkpoint
  gates, both under a cgroup cap reproducing the 180 GB class
  (`systemd-run --scope -p MemoryMax=150G -p MemorySwapMax=0`)
- peak RSS recorded both ways
- tiny-fixture gate `qwen4exp_gpu_gate` stays PASS failures=0
- per-bank byte-identity old vs new

**Comment by avifenesh** [comment]
CLAIMED q4fn 2026-09-01T21:36:26Z branch lane/q4e-loader-stream-20260901

**Comment by avifenesh** [comment]
Landed on main as 24d775458b26ca500ad7491c0f8605b33f6c99f1 (PR #52).

Peak host anon on the real artifact: **185.8 GiB -> 118.9 GiB** (199.5 GB -> 127.7 GB), measured under `systemd-run --scope -p MemoryMax=... -p MemorySwapMax=0`. The pre-streaming loader is cgroup OOM-killed mid-load at a 150 GiB cap (`anon-rss:156645208kB`); the streamed loader loads and gates clean with ~52 GB of margin on a 180 GB box.

Bytes unchanged: 63/63 bank projections byte-identical to goldens minted from the pre-streaming loader (five arms incl. per-expert modelopt at E=12), and on the real artifact an identical probe-logits sha256 `f6634e4489b1d586...` plus identical envelope table and greedy chains vs a pre-streaming reference run from an unrelated lane.

Receipts: `research/qwen4exp-bringup-20260829/loader/LOADER-STREAM.md`.

One changed consumer is written SKIPPED, not PASS: `build_tp2_shard` needs two engines and this lane held one card. That arm is the first thing to run when TP2 is next.
